front/lib/actions/mcp.ts

```diff
@@ -659,14 +661,48 @@ export class MCPConfigurationServerRunner extends BaseActionConfigurationServerR
       } else if (event.type === "notification") {
         const { notification } = event;
         if (isMCPProgressNotificationType(notification)) {
-          yield {
-            type: "tool_notification",
-            created: Date.now(),
-            configurationId: agentConfiguration.sId,
-            messageId: agentMessage.sId,
-            action: mcpAction,
-            notification: notification.params,
-          };
+          const { output: notificationOutput } = notification.params.data;
+          // Tool approval notifications have a specific handling:
+          // they are not yielded as regular notifications but are bubbled up as
+          // `tool_approve_execution` events instead, which exposes them to the end-user.
+          if (isToolApproveExecutionNotificationType(notificationOutput)) {
+            const {
+              conversationId,
+              messageId,
+              configurationId,
+              actionId,
+              inputs,
+              stake,
+              metadata,
+            } = notificationOutput.resource;
+
+            yield {
+              created: Date.now(),
+              type: "tool_approve_execution",
+              configurationId,
+              conversationId,
+              messageId,
+              actionId,
+              inputs,
+              stake,
+              metadata: {
+                ...metadata,
+                // pubsubMessageId is used to route the event to the main conversation channel.
+                pubsubMessageId: agentMessage.sId,
```

Comment:
@spolu I think you said we want to have this inside the metadata but I'm not 100% sure + should this come from run_agent 👶?

Reply:
That looks great :+1:

Reply:
Did you confirm you had an infinite loop ooc?